### PR TITLE
Modified KafkaIO.Read SDF->Legacy forced override to fail if configured functionality would be lost

### PR DESCRIPTION
**Pull request by nbali**
**Please** add a meaningful description for your change here
https://github.com/apache/beam/pull/15951 was the original trigger: basically the `DataflowRunner` replaced my SDF-based `withStopReadTime` configured `KafkaIO.Read` with a legacy read that completely lacked support for that configuration.

IMO there was two way to fix this. Either make the replacement "smart" to avoid doing it/fail-fast if it would mean a lost functionality, or completely get rid of the responsible `PTransformOverride`. So I started digging why is it needed, and why is it there. I might be missing something, but hear me out.

**2021-04-13**
So `KafkaIO.Read.KAFKA_READ_OVERRIDE` was introduced to multiple runners in https://github.com/apache/beam/commit/f2cc92663ad8ae685183e076cdb652d8fc3ba4e0 - the reasons aren't completely clear for me, and looking at the current code I'm even more baffled, but lets keep going.

**2021-05-19**
https://github.com/apache/beam/commit/99b1a926decaf372e763ea409e1d42703dda59c8 removed the `KafkaIO.Read.KAFKA_READ_OVERRIDE` from various Spark runners and instead added experiments to "*Populate experiments directly to have Kafka use legacy read.*". This seems logical to me, although - yet again - I have no idea why `DataflowRunner` wasn't touched.

**2021-05-20**
We arrived at https://github.com/apache/beam/commit/c097d8424678813901ac46f2fe674cfe71e67430 which again did some changes in the previously already modified runners. Instead of having the code responsible for the replacement in every runner, it was moved to `KafkaIO`. Yet again `DataflowRunner` was still excluded, but it had a reason now: "*Only Dataflow runner requires sdf read at this moment. For other non-portable runners, if it doesn't specify use_sdf_read, it will use legacy read regarding to performance concern.*"

------------------------

So assuming the comment from the last commit - which is still there verbatim even today - is accurate what this tells me at this point is, that with `DataflowRunner` it's not only **supported**, but actually **preferred/required** to have SDF read.

Also the comment introduced in the first commit for the `KafkaIO.Read.KAFKA_READ_OVERRIDE` is also still there and unchanged, saying: "*A `PTransformOverride` for runners to swap `ReadFromKafkaViaSDF` to legacy Kafka read if runners doesn't have a good support on executing unbounded Splittable DoFn.*" What this tells me that it should be used for runners **not supporting** SDF read.

Based on these two statements I would say it's safe to say `DataflowRunner` should **NOT** have it **AT ALL**. This PR intends to solve this problem. If you know a scenario when it's actually required I'm okay with modifying the replacement to replace it only in special scenarios, but right now I couldn't find any that would require something like that.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - ~~[ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).~~

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/validate-runner/README.md)

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by nbali** [comment]
R: @pabloem

**Comment by nbali** [comment]
R: @lukecwik

**Comment by nbali** [comment]
(I get 502 Proxy Error when trying to check what is wrong with the *Java ("Run Java PreCommit")*)

**Comment by jobegrabber** [comment]
Hi @nbali ,
if I understand correctly, exactly this override seems to be our only escape hatch to circumvent triggering [BEAM-13924](https://issues.apache.org/jira/browse/BEAM-13924), a bug that loses key/value coder information while setting up the KafkaIO.Read SDF. As you seem do have deeper understanding about the KafkaIO internals, would it be possible for you to have a look at the issue and confirm my assumption?

I'm also posting this comment for visibility as it of course doesn't invalidate any motivation for this PR.

After merging this PR we would be stuck on 2.36.0 until the linked issue is fixed.
The fix for the linked bug might be simple, I'm just now going to start looking at one.

**Comment by nbali** [comment]
I doubt this would be your workaround for that issue.
* If the problem is the SDF implementation and how differently it sets up the coders, then you can manually enforce the KafkaIO to use the legacy read. You don't need this override for that. Just pick an experiment flag to use:
https://github.com/apache/beam/blob/31332b8f0c39e94f4364a5ec1148c6450c5e78f4/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java#L1307-L1310
* If both implementation (legacy and SDF) has got issues with coders, then this replacement shouldn't matter. Actually exactly the same coder that was used before should be passed when the override this PR is trying to get rid of happens. So I don't see how having/not having this override messes with your coders if you provided them with the "intended" way.

I think it would be beneficial to copy the code snippet where your KafkaIO.Read transformations is being configured to see how exactly are you providing serializers/coders.

You can check yourself how it keeps the same coder during the override:
https://github.com/apache/beam/blob/81f3a16bccce6b2a5cb0978e71856ea53eb3bbde/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java#L1342-L1348

... and these coders are being calculated at:
https://github.com/apache/beam/blob/81f3a16bccce6b2a5cb0978e71856ea53eb3bbde/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java#L1511-L1521

Are you using the proper methods to set these up (`withKeyDeserializer`/`withKeyDeserializerAndCoder`/etc)?

**Comment by kennknowles** [comment]
Just noting that the override code path is only for runner v1, which has less mature support for the needed features. For runner v2 or prime this particular problem should not occur.

**Comment by nbali** [comment]
@kennknowles True, but it's still not the default, so many devs might encounter this.

**Comment by aaltay** [comment]
@nbali - What is the next step on this PR?

**Comment by nbali** [comment]
@aaltay I'm not aware of anything I should be doing with this.

**Comment by aaltay** [comment]
@kennknowles - were you reviewing this PR?

**Comment by kennknowles** [comment]
Yes, I am reviewing. The SDF path is supported/preferred only on runner v2 so the action for a user is to opt in to runner v2 or set any necessary experiments to use it on runner v1.

**Comment by nbali** [comment]
@kennknowles IMO v2 is either the only one supported or it's only preferred, but that two option is not the same. I'm fine either way. If v1 should forbid using the SDF version completely then I can make changes to make it fail if it's being used. Everything is better than this silent replacement.

**Comment by aaltay** [comment]
What is next on this PR?

**Comment by nbali** [comment]
@aaltay Well either approval from @kennknowles that it is okay as it is - in that case I will be resolving the conflict with the master branch -, or an instruction that I should completely forbid using SDF with v1 - which might be a braking change for some, but might correlate better with future plans.

**Comment by aaltay** [comment]
Ack. Thank you @nbali. @kennknowles might be OOO this week, he might not be able to review it this week.

**Comment by nbali** [comment]
Any ETA?

**Comment by kennknowles** [comment]
I know that @johnjcasey has looked into KafkaIO dynamic reads a bit lately. I do not know whether runner v1 has sufficient support for SDF to execute KafkaIO properly without the override. @lukecwik or @chamikaramj may also know this.

**Comment by johnjcasey** [comment]
I don't have much knowledge of whether runner can support SDF or not, sorry. I've been primarily focused on improving the SDF implementation, and less on where it has been running

**Comment by nbali** [comment]
It seemed to work, but I have to admit I only tested my use-cases and I did not perform a thorough regression test. Anyway lets assume the worst so the replacement was there for a reason, and there are cases where it doesn't work properly. In that case as I mentioned in https://github.com/apache/beam/pull/16888#issuecomment-1067548642 I can completely do the opposite and instead of the current "silently replacing SDF with Legacy" and the proposed "not overriding the SDF version with the Legacy az all", it could throw an exception if SDF implementation is being picked with v1 runner. Trully doesn't matter IMO. Just please avoid replacing it quietly as that is such a debugging nightmare.

**Comment by chamikaramj** [comment]
Is there an issue you are trying to fix here (I don't see a bug attached) ? If this was just a refactoring effort, I would just close this. Unfortunately, Kafka+Dataflow testing is internal to Google currently, so we might not figure out failures until sometime after a PR is merged.

**Comment by nbali** [comment]
@chamikaramj fully explained the the very first comment/description

**Comment by nbali** [comment]
tl;dr: v1 runner replaces sdf read with legacy read silently which is a _very bad_ user experience - it should be decided if sdf configuration with v1 runner should fail during the graph creation, or if it should be allowed and let the user handle the upcoming issues - this middle state is a pita - I went with one direction, but I don't care which is the final one, I'm just waiting for someone to say it out EXPLICITLY from the beam team which one to do...  the current middle one is bad

**Comment by chamikaramj** [comment]
I think we wanted to not fork SDK code based on the Dataflow Runner version. Hence the reason for the override in DataflowRunner.java (which is not strictly SDK). @kennknowles  has more context on this refactoring.

I think we want to keep the non-SDF version for Runner v1 since SDF on Runner v1 does not support some Dataflow features appropriately.

These I believe were the reasons for the current strange way of using SDF in the SDK and overriding for Runner v1. I don't think either removing the override or forking the SDK is appropriate. May be we can update Java docs or Beam Website documentation to explain this better ?

**Comment by kennknowles** [comment]
I guess I'd say it the opposite way: Dataflow runner v1 does not support SDF appropriately.

I don't think it is strange to use SDF in the SDK and override in DataflowRunner. This is exactly what all runners do: they translate Beam's primitives to something they can execute. Runners are allowed and encouraged to directly translate composites as long as they preserve the semantics.

So my summary take is still:

 - Dataflow runner v1 doesn't support SDF very well, and I don't really know the boundaries of where the support actually works vs where it crashes vs where it fails silently.
 - Dataflow runner v1 can still execute Kafka reads by executing the composite directly in a custom way, and that sort of thing is explicitly a design goal of Beam.
 - If this results in mysterious behavior or inscrutable error messages, that is a bug.
 - If we can positively confirm (via continuously running tests) that Dataflow runner v1 correctly supports some functionality, then we can enable it.

So I understand the motivation here, but I just don't think it is safe.

**Comment by nbali** [comment]
Actually now that https://github.com/apache/beam/pull/16909 has been finally merged, I think we don't even need this anymore.

The override replaces `ReadFromKafkaViaSDF` with `ReadFromKafkaViaUnbounded` by calling the constructor of `ReadFromKafkaViaUnbounded`.
https://github.com/apache/beam/blob/c6972f4b8d9a03f4639395f60f18e740dce91398/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java#L1345-L1371

The constructor of `ReadFromKafkaViaUnbounded`
https://github.com/apache/beam/blob/c6972f4b8d9a03f4639395f60f18e740dce91398/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java#L1392-L1395
 will call `AbstractReadFromKafka`
https://github.com/apache/beam/blob/c6972f4b8d9a03f4639395f60f18e740dce91398/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java#L1373-L1385
which contains `KafkaIOReadImplementationCompatibility.getCompatibility(kafkaRead).checkSupport(implementation);`
https://github.com/apache/beam/blob/c6972f4b8d9a03f4639395f60f18e740dce91398/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIOReadImplementationCompatibility.java#L232-L240
which should fail if we try to initiate the DoFn with a property that is not supported.

So I think we either get the legacy read without a feature loss, or we fail during the graph building if we would have a feature loss.

Is that acceptable then? @kennknowles  @chamikaramj

**Comment by nbali** [comment]
> Runners are allowed and encouraged to directly translate composites as long as they preserve the semantics.

I think the previously described behaviour is exactly this now. (Previously semantics was lost.)

**Comment by nbali** [comment]
I mean it will be still a bit confusing that someone tries to use SDF, and gets an error msg that LEGACY doesn't support that property, but it's still better than previously. I could add try-catch and re-wrap the exception with the proper msg in the override if needed.

**Comment by kennknowles** [comment]
Interesting. We certainly shouldn't be calling anything "legacy" when it is the default. I know some people like to imagine that runner v2 and portability are "current" but actually they are new and opt-in.

**Comment by kennknowles** [comment]
If a KafkaIO feature is used that default runner v1 does not support, the error message should say "this feature is only supported by Dataflow runner v2, try --dataflowServiceOptions=use_runner_v2".

**Comment by nbali** [comment]
@kennknowles Okay, I have modified the PR as requested. Although I modified the suggested error message a bit as this is a general feature. It's just the current situation that only the DF runner is the only one using it.

**Comment by aaltay** [comment]
@kennknowles - would you be able to review this PR?

**Comment by aaltay** [comment]
@kennknowles - Could someone else help you with reviewing this change?

**Comment by nbali** [comment]
ping?

**Comment by kennknowles** [comment]
I have not been able to follow all the changes. It looks like @TheNeuralBit merged the other PR? This one seems like a very small tweak that basically depends on the semantics of that one.

**Comment by aaltay** [comment]
> I have not been able to follow all the changes. It looks like @TheNeuralBit merged the other PR? This one seems like a very small tweak that basically depends on the semantics of that one.

Could @TheNeuralBit review this?

**Review by TheNeuralBit — APPROVED** [review]

**Comment by TheNeuralBit** [comment]
retest this please

**Comment by nbali** [comment]
@TheNeuralBit success

**Comment by aaltay** [comment]
Merging based on the passing tests and @TheNeuralBit 's approval.

**Comment by aaltay** [comment]
Run Java_Kafka_IO_Direct PreCommit

**Comment by aaltay** [comment]
> Merging based on the passing tests and @TheNeuralBit 's approval.

Never mind. There was a pending test.

**Comment by nbali** [comment]
I have to admin, I'm pretty confused right now.

**Comment by aaltay** [comment]
Merging for real, all tests are passing.
